src/activity_actor.cpp

```diff
@@ -8435,93 +8432,99 @@ void pulp_activity_actor::do_turn( player_activity &act, Character &you )
-                //don't smash acid zombies when auto pulping unprotected
-                continue;
-            }
-            while( corpse.damage() < corpse.max_damage() ) {
-                // Increase damage as we keep smashing ensuring we eventually smash the target.
-                if( x_in_y( pulp_power, corpse.volume() / 250_ml ) ) {
-                    corpse.inc_damage();
-                    if( corpse.damage() == corpse.max_damage() ) {
-                        num_corpses++;
-                    }
-                }
+    int map_stack_counter = 0;
+    tripoint_bub_ms pos = here.get_bub( *current_pos_iter );
+    map_stack corpse_pile = here.i_at( pos );
+    for( item &corpse : corpse_pile ) {
+        if( !corpse.is_corpse() || !corpse.can_revive() ) {
+            // Don't smash non-rezing corpses or random items
+            map_stack_counter++;
+            continue;
+        }
 
-                if( x_in_y( pulp_power, corpse.volume() / 250_ml ) ) {
-                    // Splatter some blood around
-                    // Splatter a bit more randomly, so that it looks cooler
-                    const int radius = mess_radius + x_in_y( pulp_power, 500 ) + x_in_y( pulp_power, 1000 );
-                    const tripoint_bub_ms dest( pos + point( rng( -radius, radius ), rng( -radius, radius ) ) );
-                    const field_type_id type_blood = ( mess_radius > 1 && x_in_y( pulp_power, 10000 ) ) ?
-                                                     corpse.get_mtype()->gibType() :
-                                                     corpse.get_mtype()->bloodType();
-                    here.add_splatter_trail( type_blood, pos, dest );
+        const mtype *corpse_mtype = corpse.get_mtype();
+        const bool acid_immune = you.is_immune_damage( damage_acid ) ||
+                                 you.is_immune_field( fd_acid );
+        // this corpse is acid, and you are not immune to it
+        bool acid_corpse = corpse_mtype->bloodType().obj().has_acid && !acid_immune;
+
+        if( corpse.damage() < corpse.max_damage() ) {
+            // Increase damage as we keep smashing ensuring we eventually smash the target.
+            // Slow down the speed if corpse is acid
+            if( x_in_y( pulp_power, corpse.volume() / ( acid_corpse ? 20_ml : 50_ml ) ) ) {
+                corpse.inc_damage();
+                if( corpse.damage() == corpse.max_damage() ) {
+                    num_corpses++;
+                    corpse.set_flag( flag_PULPED );
                 }
+            }
 
-                // mixture of isaac clarke stomps and swinging your weapon
-                you.burn_energy_all( -pulp_effort );
-                you.recoil = MAX_RECOIL;
-
-                if( one_in( 4 ) ) {
-                    // Smashing may not be butchery, but it involves some zombie anatomy
-                    you.practice( skill_survival, 2, 2 );
+            if( x_in_y( pulp_power, corpse.volume() / 50_ml ) ) {
```

Comment:
it's surprising that this is a separate rng check. Is it, like, good? Should it not be folded into the above one?

Reply:
it can be, yes